### ISSUE TITLE
With Android Studio, classes are locked on Windows

### ISSUE THREAD

**Issue by thierryd** [jshiell/checkstyle-idea#263]
I'm using Android Studio 2.2 and the checkstyle-idea plugin locks the classes.jar files. If I start Android Studio clean and do a full Instant Run build, the gradle process create the APK. After the BUILD SUCCESSFUL appears, if I use Process Explorer (https://technet.microsoft.com/en-us/sysinternals/processexplorer.aspx), the classes.jar files are locked.

I initially thought that this was an Android Studio issue and I've entered my detail `[here](https://code.google.com/p/android/issues/detail?id=202297)`.

With the help of the folks at Google, I was able to see the stacktrace for the locked classes.  For example, for the "classes.jar" of the multidex library:

```
#14 C:\dev\workspace\android_replica\android_replica_app\app\build\intermediates\exploded-aar\com.android.support\multidex-instrumentation\1.0.1\jars\classes.jar by thread:Thread-25 on Mon Aug 01 10:55:36 EDT 2016
    at java.util.zip.ZipFile.<init>(ZipFile.java:150)
    at java.util.jar.JarFile.<init>(JarFile.java:166)
    at java.util.jar.JarFile.<init>(JarFile.java:103)
    at sun.misc.URLClassPath$JarLoader.getJarFile(URLClassPath.java:893)
    at sun.misc.URLClassPath$JarLoader.access$700(URLClassPath.java:756)
    at sun.misc.URLClassPath$JarLoader$1.run(URLClassPath.java:838)
    at sun.misc.URLClassPath$JarLoader$1.run(URLClassPath.java:831)
    at java.security.AccessController.doPrivileged(Native Method)
    at sun.misc.URLClassPath$JarLoader.ensureOpen(URLClassPath.java:830)
    at sun.misc.URLClassPath$JarLoader.<init>(URLClassPath.java:803)
    at sun.misc.URLClassPath$3.run(URLClassPath.java:530)
    at sun.misc.URLClassPath$3.run(URLClassPath.java:520)
    at java.security.AccessController.doPrivileged(Native Method)
    at sun.misc.URLClassPath.getLoader(URLClassPath.java:519)
    at sun.misc.URLClassPath.getLoader(URLClassPath.java:492)
    at sun.misc.URLClassPath.getNextLoader(URLClassPath.java:457)
    at sun.misc.URLClassPath.getResource(URLClassPath.java:211)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:365)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:348)
    at com.puppycrawl.tools.checkstyle.PackageObjectFactory.createObject(PackageObjectFactory.java:155)
    at com.puppycrawl.tools.checkstyle.PackageObjectFactory.createObjectWithIgnoringProblems(PackageObjectFactory.java:114)
    at com.puppycrawl.tools.checkstyle.PackageObjectFactory.createModule(PackageObjectFactory.java:87)
    at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:397)
    at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:138)
    at org.infernus.idea.checkstyle.checker.CheckerFactoryWorker.createCreater(CheckerFactoryWorker.java:63)
    at org.infernus.idea.checkstyle.checker.CheckerFactoryWorker.run(CheckerFactoryWorker.java:51)
```

This is really frustrating because when Android Studio detect a new device connected, a "gradle clean" command is executed, which fail because it cannot delete the build folder.

**Comment by jshiell** [comment]
Thanks for the report.

This is probably a side effect of the retention of Checkers for short periods (I think about 30s) so as to avoid constantly recreating them. It might be interesting to see if the refresh action (which is basically a cache clear) would resolve this - if so, one possibility would be seeing if there was a way to link that into build clear events (mind, I'm only guessing that such things would exist, but it'd seem a sensible extension point).

In the absence of such, this becomes a harder nut to crack. Apart from anything else, I suspect this won't affect *nix users, given the lesser restrictions on locked files, and I'm currently using OS X development.

**Comment by madisparn** [comment]
Also happens on Intellij Ultimate

**Comment by henri5** [comment]
I'm experiencing the same issue as @madisparn , is there a way to have it ignore .jars (why is it checking those anyway?).

**Comment by jshiell** [comment]
> (why is it checking those anyway?).

Because Checkstyle needs access to the classpath for the exception checks, among others.

Previous point still stands I fear - I don't use Windows, so if a fix does come then it's likely to be by accident.

**Comment by tsjensen** [comment]
We would really need to know if the problem persists after you press the little "recycle" icon in the Checkstyle view (the popup text says "Reload Rules Files"). This would affect possible solutions.

@madisparn I tried it on IntelliJ Ultimate, too, but could not reproduce the problem. Do you have a sample project / step by step instructions we could follow to reproduce?

If the hypothetical "clean project hook" does not exist, we might want to offer a new option in the plugin configuration to throw away the checkers every time. This would slow down real-time inspections, but may resolve this problem.

@thierryd Are you sure you are not somehow executing Checkstyle inspections and the clean build at the same time. So, all editor windows closed, no relevant background process running, etc. Because if so, the solution becomes even more complex, as we'd have to notice the situation and cancel running inspections.

**Comment by tsjensen** [comment]
In @thierryd's linked [Android issue](https://code.google.com/p/android/issues/detail?id=202297), it is suggested that IDEA has a feature that copies libraries to prevent their getting locked. That's radical, but might also be worth exploring as an option.

**Comment by henri5** [comment]
Attached sample project to reproduce the issue.

1. Project is pre-built, open it in IntelliJ, verify that `lib/jsr305.jar` is added to module dependencies
2. run `mvn clean package` - cleans nicely
3. In checkstyle panel select _Google checks_ and scan project
4. run `mvn clean package` again - fails to delete `jsr305.jar`

Pressing _Reload Rules files_ button does not help, the 'copy library' feature does not help either, being enabled on Windos by default anyway:

> By default this behavior is enabled on Windows and disabled on other platforms.

[checkstyle-win-lock.zip](https://github.com/jshiell/checkstyle-idea/files/1487002/checkstyle-win-lock.zip)

**Comment by tsjensen** [comment]
Thanks @henri5, this helps a lot. I can reproduce the problem with IntelliJ IDEA CE on Windows using your sample project.

@jshiell What are your plans? I can offer to give it a shot, but as always it will take a lot of patience.

I would pursue a selective copy approach, by which we only copy libraries from inside the project directory. I couldn't wrap my mind around the build hook idea, especially when the build is executed from the command line. We would re-copy only when the library's timestamp or size has changed.

**Comment by jshiell** [comment]
@tsjensen I fear my plans are non-existent at present, as I don't have a Windows dev env (I'm using OS X at home, and OS X/Linux at work). Any effort you can spare on this would be (as always) gratefully received!

**Comment by tsjensen** [comment]
PR is here. Let's hope this will do the trick.

**Comment by jshiell** [comment]
Merged and released as 5.14.0. Thanks yet again to the wonderful @tsjensen 😄
